File: 2020/7/main.hs
import qualified Data.Set as S
import Debug.Trace (traceShow)
import Text.Regex.Posix ((=~))

data Bag = Bag {name :: String, count :: Int}
  deriving (Show, Ord, Eq)

contains :: String -> [Bag] -> Maybe Bag
contains key bags = head $ filter (\x -> name x == key) bags

parseLine :: String -> (String, [Bag])
parseLine line = (bagName, subBags)
  where
    bagName = line =~ "[a-z]+ [a-z]+" :: String
    subMatches = line =~ "([0-9]+) ([a-z]+ [a-z]+)" :: [[String]]
    subBags = [Bag a (read b :: Int) | [a, b, _] <- subMatches]

solve1 :: [(String, [Bag])] -> String -> S.Set String -> S.Set String
solve1 [] _ accu = accu
solve1 (b : allBags) key accu = solve1 allBags key newSet
  where
    match = contains key (snd b)
    newSet = S.insert (name match) accu

main = do
  lines <- fmap lines (readFile "test.txt")
  let x = map parseLine lines
  print $ take 5 x
  print $ solve1 x "shiny gold" S.empty

File: 2020/7/main.py
import re

with open("test.txt") as f:
    bags = {}
    for line in f.readlines():
        key = re.findall(r"[a-z]+ [a-z]+", line)[0]
        bags[key] = {}

        subs = re.findall(r"([1-9]+) ([a-z]+ [a-z]+)", line)
        for count, sub in subs:
            bags[key][sub] = int(count)

print(bags)


def find(key, contained_in):
    for bag, contained in bags.items():
        if key in contained:
            contained_in.add(bag)
            find(bag, contained_in)


def find2(key):
    total = 1
    for bag_key, count in bags[key].items():
        total += count * find2(bag_key)

    return total


contained_in = set()
find("shiny gold", contained_in)
print(len(contained_in))

x = find2("shiny gold")
print(x - 1)
